### PR TITLE
[KBFS-1024] Unembed BareTlfHandle from TlfHandle

### PR DESCRIPTION
**Pull request by akalin-keybase**
This will let TlfHandle evolve separately from
BareTlfHandle. In particular, this will let
TlfHandle keep track of per-assertion info
without duplicating info.

Beef up BareTlfHandle and TlfHandle tests.

### CODE REVIEW

File: libkbfs/bare_tlf_handle.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+	si := u[i].String()
+	sj := u[j].String()
+	return si < sj
+}
+
+func (u socialAssertionList) Swap(i, j int) {
+	u[i], u[j] = u[j], u[i]
+}
+
+// MakeBareTlfHandle creates a BareTlfHandle from the given list of
```
Comment (akalin-keybase) [inline libkbfs/bare_tlf_handle.go]: Added more comments here

File: libkbfs/bare_tlf_handle.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
+package libkbfs
```
Comment (akalin-keybase) [inline libkbfs/bare_tlf_handle.go]: This file is just code moved from tlf_handle.go, except when annotated below

File: libkbfs/bare_tlf_handle.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+	return h.findUserInList(user, h.Writers)
+}
+
+// IsReader returns whether or not the given user is a reader for the
+// top-level folder represented by this BareTlfHandle.
+func (h BareTlfHandle) IsReader(user keybase1.UID) bool {
+	return h.IsPublic() || h.findUserInList(user, h.Readers) || h.IsWriter(user)
+}
+
+// ResolvedUsers returns the concatenation of h.Writers and h.Readers,
```
Comment (akalin-keybase) [inline libkbfs/bare_tlf_handle.go]: Changed the behavior of this function slightly (to not return PUBLIC_UID)

File: libkbfs/bare_tlf_handle.go [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+	resolvedUsers = append(resolvedUsers, h.Writers...)
+	if !h.IsPublic() {
+		resolvedUsers = append(resolvedUsers, h.Readers...)
+	}
+	return resolvedUsers
+}
+
+// HasUnresolvedUsers returns true if this handle has any unresolved
+// writers or readers.
+func (h BareTlfHandle) HasUnresolvedUsers() bool {
```
Comment (akalin-keybase) [inline libkbfs/bare_tlf_handle.go]: Added this helper

File: libkbfs/bare_tlf_handle.go [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+	return len(h.UnresolvedWriters) > 0 || len(h.UnresolvedReaders) > 0
+}
+
+// UnresolvedUsers returns the concatenation of h.UnresolvedWriters
```
Comment (akalin-keybase) [inline libkbfs/bare_tlf_handle.go]: Specified order in comment

File: libkbfs/interfaces.go [hunk 6]
```diff
@@ -738,7 +738,7 @@ type MDOps interface {
 	// which may not yet be reflected in the MD if the TLF hasn't been rekeyed since it
 	// entered into a conflicting state.
 	GetLatestHandleForTLF(ctx context.Context, id TlfID) (
-		*BareTlfHandle, error)
+		BareTlfHandle, error)
```
Comment (akalin-keybase) [inline libkbfs/interfaces.go]: There wasn't any special reason for this to return a pointer, right?

File: libkbfs/key_manager.go [hunk 7]
```diff
@@ -446,24 +446,22 @@ func (km *KeyManagerStandard) Rekey(ctx context.Context, md *RootMetadata, promp
 		if err != nil {
 			return false, nil, err
 		}
-		if latestHandle == nil {
-			return false, nil, NoSuchTlfHandleError{md.ID}
-		}
-		if latestHandle.ConflictInfo != resolvedHandle.ConflictInfo {
+		resolvedInfo := resolvedHandle.ConflictInfo()
+		if latestHandle.ConflictInfo != resolvedInfo {
```
Comment (akalin-keybase) [inline libkbfs/key_manager.go]: This is a pointer comparison. Is this really correct? I would have expected this to use `CodecEqual()`. @aalness

File: libkbfs/bare_tlf_handle.go (hunk 3 above)
Comment (strib) [inline libkbfs/bare_tlf_handle.go]: Looks like the caller of the existing `Users()` function, on the server side, adds an entry to the folder resolver table including the public UID.  Do you plan to have a server-side PR that replaces that behavior?  Or is it not needed / relied on?

File: libkbfs/interfaces.go (hunk 6 above)
Comment (strib) [inline libkbfs/interfaces.go]: I think it was just to be consistent with the other interface functions that take `*TlfHandle`.  Probably no real reason.

File: libkbfs/key_manager.go (hunk 7 above)
Comment (strib) [inline libkbfs/key_manager.go]: Good catch, I think you're right, it should be `CodecEqual()`.

File: libkbfs/tlf_handle.go [hunk 8]
```diff
@@ -16,246 +16,157 @@ import (
...
-
-func (u UIDList) Less(i, j int) bool {
-	return u[i].Less(u[j])
-}
+// CanonicalTlfName is a string containing the canonical name of a TLF.
+type CanonicalTlfName string
 
-func (u UIDList) Swap(i, j int) {
-	u[i], u[j] = u[j], u[i]
+// TlfHandle is BareTlfHandle augmented with additional info.
```
Comment (strib) [inline libkbfs/tlf_handle.go]: Maybe explain why a BH is not just embedded?

File: vendor/github.com/keybase/client/go/auth/credential_authority.go [hunk 9]
```diff
@@ -2,11 +2,12 @@ package auth
 
 import (
 	"fmt"
+	"time"
```
Comment (strib) [inline vendor/github.com/keybase/client/go/auth/credential_authority.go]: No `vendor.json` update?  Is this supposed to be here?  Is there a corresponding `client` PR?

**Comment by strib** [comment]
Looks good to me, mod comments.  Thanks!

File: libkbfs/bare_tlf_handle.go (hunk 3 above)
Comment (akalin-keybase) [inline libkbfs/bare_tlf_handle.go]: I'll have a server-side PR, but I don't see where it puts an entry. Let's discuss it on that PR...

File: libkbfs/key_manager.go (hunk 7 above)
Comment (akalin-keybase) [inline libkbfs/key_manager.go]: Fixed.

File: libkbfs/tlf_handle.go (hunk 8 above)
Comment (akalin-keybase) [inline libkbfs/tlf_handle.go]: Done.

File: vendor/github.com/keybase/client/go/auth/credential_authority.go (hunk 9 above)
Comment (akalin-keybase) [inline vendor/github.com/keybase/client/go/auth/credential_authority.go]: Oops, removed.

**Pull request merged**